Lib/concurrent/futures/_base.py

```diff
@@ -544,18 +545,40 @@ def map(self, fn, *iterables, timeout=None, chunksize=1):
         """
         if timeout is not None:
             end_time = timeout + time.time()
+        if prefetch is None:
+            prefetch = self._max_workers
+        if prefetch < 0:
+            raise ValueError("prefetch count may not be negative")
 
-        fs = [self.submit(fn, *args) for args in zip(*iterables)]
+        argsiter = zip(*iterables)
+
+        fs = collections.deque(self.submit(fn, *args) for args in itertools.islice(argsiter, self._max_workers + prefetch))
 
         # Yield must be hidden in closure so that the futures are submitted
         # before the first iterator value is required.
         def result_iterator():
+            nonlocal argsiter
             try:
-                for future in fs:
+                while fs:
                     if timeout is None:
-                        yield future.result()
+                        res = fs[0].result()
                     else:
-                        yield future.result(end_time - time.time())
+                        res = fs[0].result(end_time - time.time())
+
+                    # Got a result, future needn't be cancelled
+                    del fs[0]
+
```

Comment:
I believe you can't safely use popleft due to the timeout case; if the timeout occurs, you need to explicitly cancel the future, so it needs to stay in the deque until we're sure it finished, one way or the other. I could popleft, then have a dedicated try/finally to manage it, but I felt it was simpler to delay deletion and let the top level try/finally cleanup handle it. As a note, del fs[0] does internally make a couple unnecessary C function calls (rotating it by 0, popleft, then rotating back by 0, though the rotates by 0 are largely no-ops), but in practice, the overhead of generic function calls for a Python level call to popleft that syntax based deletion avoids means it's all likely a wash.